File: onchain_registry/ipfs_test.js
const fetch = require('node-fetch');
const FormData = require('form-data');

const IPFS_API_URL = 'http://127.0.0.1:5001/api/v0';

async function uploadAndPin(jsonData) {
    try {
        const formData = new FormData();
        const jsonBuffer = Buffer.from(JSON.stringify(jsonData));
        formData.append('file', jsonBuffer, { filename: 'data.json', contentType: 'application/json' });
        
        const response = await fetch(`${IPFS_API_URL}/add`, {
            method: 'POST',
            body: formData
        });
        
        if (!response.ok) throw new Error(`HTTP error! status: ${response.status}`);
        
        const data = await response.json();
        console.log(`Uploaded JSON and received CID: ${data.Hash}`);
        return data.Hash;
    } catch (error) {
        console.error(`Failed to upload JSON: ${error.message}`);
        return null;
    }
}

async function deleteAndUnpin(cid) {
    try {
        const response = await fetch(`${IPFS_API_URL}/pin/rm?arg=${cid}`, {
            method: 'POST'
        });
        
        if (!response.ok) throw new Error(`HTTP error! status: ${response.status}`);
        
        console.log(`Unpinned CID: ${cid}`);
    } catch (error) {
        console.error(`Failed to unpin CID: ${error.message}`);
    }
}

async function getFile(cid) {
    try {
        const response = await fetch(`${IPFS_API_URL}/cat?arg=${cid}`, {
            method: 'POST'
        });
        
        if (!response.ok) throw new Error(`HTTP error! status: ${response.status}`);
        
        const jsonData = await response.json();
        console.log(`Retrieved JSON data:`, jsonData);
        return jsonData;
    } catch (error) {
        console.error(`Failed to retrieve JSON: ${error.message}`);
        return null;
    }
}

module.exports = {
    uploadAndPin,
    deleteAndUnpin,
    getFile
};

// Example usage
if (require.main === module) {
    const dataToUpload = {
        message: "Hello, IPFS!",
        timestamp: "2025-01-21",
        example: true
    };

    (async () => {
        try {
            const cid = await uploadAndPin(dataToUpload);
            if (cid) {
                const retrievedData = await getFile(cid);
                await deleteAndUnpin(cid);
            }
        } catch (error) {
            console.error(error);
        }
    })();
}

File: onchain_registry/ipfs_test.py
import requests
import json

IPFS_API_URL = "http://127.0.0.1:5001/api/v0"

def upload_and_pin(json_data):
    # Convert JSON data to a string and upload it
    response = requests.post(f"{IPFS_API_URL}/add", files={"file": ("data.json", json.dumps(json_data))})
    if response.status_code == 200:
        cid = response.json()["Hash"]
        print(f"Uploaded JSON and received CID: {cid}")
        return cid
    else:
        print(f"Failed to upload JSON: {response.text}")
        return None

def delete_and_unpin(cid):
    response = requests.post(f"{IPFS_API_URL}/pin/rm?arg={cid}")
    if response.status_code == 200:
        print(f"Unpinned CID: {cid}")
    else:
        print(f"Failed to unpin CID: {response.text}")

def get_file(cid):
    response = requests.post(f"{IPFS_API_URL}/cat?arg={cid}")
    if response.status_code == 200:
        # Parse JSON content from IPFS
        json_data = json.loads(response.content)
        print(f"Retrieved JSON data: {json_data}")
        return json_data
    else:
        print(f"Failed to retrieve JSON: {response.text}")
        return None

if __name__ == "__main__":
    # Step 1: Define the JSON data
    data_to_upload = {
        "message": "Hello, IPFS!",
        "timestamp": "2025-01-21",
        "example": True
    }

    # Step 2: Upload and pin the JSON data
    cid = upload_and_pin(data_to_upload)

    if cid:
        # Step 3: Retrieve the JSON data
        retrieved_data = get_file(cid)

        # Step 4: Unpin and delete the JSON data
        delete_and_unpin(cid)
